src/Reflex/Backend/Socket/Connect.hs

```diff
@@ -1,91 +1,70 @@
-    _ccHostname :: Maybe String
-  , _ccPort     :: Maybe String
-  }
-
-makeLenses ''ConnectConfig
-
-data Connect t =
-  Connect {
-    _cSocket :: Event t Socket
-  , _cError  :: Event t IOException
-  }
-
-makeLenses ''Connect
-
-connect ::
-  ( Reflex t
-  , PerformEvent t m
-  , TriggerEvent t m
-  , PostBuild t m
-  , MonadIO (Performable m)
-  , MonadIO m
-  ) =>
-  ConnectConfig ->
-  m (Connect t)
-connect (ConnectConfig mHost mPort) = do
-  (eSocket, onSocket) <- newTriggerEvent
-  (eError, onError)   <- newTriggerEvent
-
-  let
-    exHandlerGetAddr :: IOException -> IO [AddrInfo]
-    exHandlerGetAddr e = [] <$ onError e
-
-    getAddr :: IO (Maybe AddrInfo)
-    getAddr = do
-      addrInfos <- getAddrInfo Nothing mHost mPort `catch` exHandlerGetAddr
-      pure $ case addrInfos of
-        [] -> Nothing
-        h : _ -> Just h
-
-    exHandler :: IOException -> IO (Maybe Socket)
-    exHandler e = Nothing <$ onError e
-
-    connectAddr :: AddrInfo -> IO (Maybe Socket)
-    connectAddr h = do
-      sock <- socket (addrFamily h) Stream defaultProtocol
-      NS.connect sock (addrAddress h)
-      pure (Just sock)
-
-    start = liftIO $ do
-      mAddrInfo <- getAddr
-      forM_ mAddrInfo $ \h -> do
-        mSock <- connectAddr h `catch` exHandler
-        forM mSock onSocket
-
+import           Reflex
+
+-- | Connect to a remote endpoint. The connection happens in a
+-- background thread.
+connect
+  :: ( Reflex t
+     , PerformEvent t m
+     , TriggerEvent t m
+     , PostBuild t m
+     , MonadIO (Performable m)
+     , MonadIO m
+     )
+  => Maybe (NS.HostName)
+     -- ^ Host to connect to. If 'Nothing', connect via loopback.
+  -> NS.ServiceName
+     -- ^ Service (port number or service name). See the
+     -- <https://linux.die.net/man/3/getaddrinfo manpage for getaddrinfo>.
+  -> m (Event t (Either IOException Socket))
+     -- ^ This event will fire exactly once.
+connect mHost service = do
   ePostBuild <- getPostBuild
-  performEvent_ $ start <$ ePostBuild
-
-  pure $ Connect eSocket eError
+  performEventAsync $ ePostBuild $> \onRes -> void . liftIO . forkIO $
+    let
+      getAddrs :: ExceptT IOException IO (NonEmpty AddrInfo)
+      getAddrs = ExceptT . try $
+        -- fromList is OK here, as getaddrinfo(3) is required to
+        -- return a nonempty list of addrinfos.
+        --
+        -- See: http://pubs.opengroup.org/onlinepubs/9699919799/functions/getaddrinfo.html
+        -- And: https://github.com/haskell/network/issues/407
+        fromList <$> NS.getAddrInfo Nothing mHost (Just service)
+
+      tryConnect :: AddrInfo -> ExceptT IOException IO Socket
+      tryConnect info = ExceptT . try $ do
+        sock <- NS.socket (addrFamily info) NS.Stream defaultProtocol
+        NS.connect sock (addrAddress info) `onException` NS.close sock
+        pure sock
+
+    in do
+      res <- runExceptT $ do
+        addrs <- getAddrs
+        let attempts = withExceptT Last . tryConnect <$> addrs
```

Comment:
Should this be consistent with the error reporting in `Accept`?

Reply:
I've changed this, but now I have two isomorphic error types and no good idea of what a unifying name might be (apart from `Either`). Thoughts?

Reply:
I called it `SetupError` and moved it into its own module.